de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/SelectedEditorState.java

```diff
@@ -0,0 +1,157 @@
+package de.fu_berlin.inf.dpp.intellij.editor;
+
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.VirtualFile;
+
+import de.fu_berlin.inf.dpp.SarosPluginContext;
+import de.fu_berlin.inf.dpp.filesystem.IFile;
+import de.fu_berlin.inf.dpp.intellij.filesystem.IntelliJProjectImplV2;
+
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+import org.picocontainer.annotations.Inject;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ListIterator;
+
+/**
+ * Class used to capture and re-apply which editors are currently selected by
+ * the user.
+ */
+//TODO consider duplicated open editors during screen splitting
+public class SelectedEditorState {
+    private static final Logger log = Logger
+        .getLogger(SelectedEditorState.class);
+
+    private final List<VirtualFile> selectedEditors;
+
+    private boolean hasCapturedState;
+
+    @Inject
+    private ProjectAPI projectAPI;
+
+    @Inject
```

Comment:
I implemented this class to use it in the upcoming rework of [SharedResourcesManager](https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/SharedResourcesManager.java) and [FileSystemChangeListener](https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/FileSystemChangeListener.java). In both classes, we would also be missing Project (and, in the upcoming change, SharedResourcesManager would be missing ProjectAPI if it didn't have to hand it to FileSystemChangeListener).

My thinking was that I wanted to keep the usage of this class as simple as possible as it is meant as a general purpose utility class. I didn't want to pass these objects through the CTOR as it too often leads to a situation where other classes wanting to use this class would themselves have to use CTOR injection and add the needed objects as instance variables just to use this class.

So in short, I did not know which objects were present anyway in the classes using this class. I could just go by the current use cases, but then passing some the needed objects and using annotation injection for others seems like a weird halve-measure to me.

I see the point that this is just wasted computing time if all the needed objects are already present, but I still think this is a cleaner solution. But if you strongly disagree, I can change it.

Reply:
The boilerplated solution would be to add a factory that has only one method. I really dislike the injection part as the PicoContainer library is discontinued since ages and there are potential deadlock pits especially when calling start and stop on such a container (as we do this not in the Application thread). This is also the reason why ISarosSession#getComponent() returns null when the session is in the stopping phase, simply to avoid deadlocking.

Reply:
Just forgot what I said, this will not help as this will result in a cyclic dependency.

Factory -> A, B, C

A -> Factory

Reply:
I am not sure what your current stance on the implementation using annotation injection is.
Is the lifecycle stuff really such a big problem, even though this is using the plugin container/context (instead of the much more short lived session context)?
If you would still like to avoid annotation injection, I will change it.

On a related note: How pressing is the whole PicoContainer being legacy/unmaintained thing in your opinion?

There are some alternatives for dependency injection libraries, but the switch will probably be very work intensive. Especially the regression testing to check that the switch didn't break anything. But, depending on the importance, this could be something we could be working on after the gradle stuff and the subsequent JDK upgrade is done.

(In general I think we should be working towards updating all our libraries and ditching deprecated libraries to create a more current and supported base state; the problem with the PicoContainer is that it is a central and complex component in our logic, which makes this process very difficult.)

Reply:
Saros.initComponent() just creates and attaches another container. It then does the IOC stuff and detaches the container. I really do not know if it is able to inject dependencies that are part of the session container. This method is not intended to be called over and over again(e.g 1000 times per second). It is more designed to offer access to the context for objects that are not instanciated by us, e.g in Saros/E we have several classes that are created by Eclipse at a specific time which is far behind the creation/activating and intializing the plugin context for our plugin.


As for the whole framework. We do not really use much stuff that PicoContainer offers. I already did take a look at Guice and came to the conclusion that a scoped context is much harder to implement with Guice. And as you already mentioned it would not be an easy task to switch the IOC framework and therefore I do not really see any benefits in doing it, but only wasting time.

Reply:
How about making the variables and the injection static? The used objects should not change during the lifetime of the plugin.

```Java
@Inject
private static ProjectAPI projectAPI;

@Inject
private static EditorManager editorManager;

@Inject
private static Project project;

static {
    SarosPluginContext.initComponent(new SelectedEditorState());
}
```

As a side note, this looks kind of confusing with annotation injection. But we don't have `getComponent(Class)` for the SarosPluginContext. I still think we could add it, as proposed in [3295](https://saros-build.imp.fu-berlin.de/gerrit/#/c/3295/). It would also remove the false positives where IntelliJ states that variables using annotation injection are never set.